test/unit/core/createOptIn.spec.js

```diff
@@ -0,0 +1,122 @@
+      cookie = jasmine.createSpyObj("cookie", ["get", "set"]);
+      cookie.get.and.returnValue(null);
+    });
+
+    it("considers the user opted in if cookie is set to 'all'", () => {
+      cookie.get.and.returnValue("all");
+      optIn.enable(cookie);
+
+      expect(optIn.isOptedIn()).toBe(true);
+      return optIn.whenOptedIn();
+    });
+
+    it("considers the user opted out if cookie is set to 'none'", done => {
+      cookie.get.and.returnValue("none");
+      optIn.enable(cookie);
+
+      expect(optIn.isOptedIn()).toBe(false);
+      optIn.whenOptedIn().catch(error => {
+        expect(error).toBeDefined();
+        done();
+      });
+    });
+
+    it("considers the user pending opt in if cookie is not set", () => {
+      const optedInSpy = jasmine.createSpy();
+      optIn.enable(cookie);
+
+      expect(optIn.isOptedIn()).toBe(false);
+      optIn.whenOptedIn().then(optedInSpy);
+      return flushPromises().then(() => {
+        expect(optedInSpy).not.toHaveBeenCalled();
+      });
+    });
+
+    it("considers the user opted in after the user opts in", () => {
+      const optedInSpy = jasmine.createSpy();
+      optIn.enable(cookie);
+      optIn.whenOptedIn().then(optedInSpy);
+      optIn.setPurposes("all");
+
+      expect(cookie.set).toHaveBeenCalledWith("optIn", "all");
+      expect(optIn.isOptedIn()).toBe(true);
+      return flushPromises().then(() => {
+        expect(optedInSpy).toHaveBeenCalled();
+      });
+    });
+
+    it("considers the user opted out after the user opts out", () => {
+      const optedOutSpy = jasmine.createSpy();
+      optIn.enable(cookie);
+      optIn.whenOptedIn().catch(optedOutSpy);
+      optIn.setPurposes("none");
+
+      expect(cookie.set).toHaveBeenCalledWith("optIn", "none");
+      expect(optIn.isOptedIn()).toBe(false);
+      return flushPromises().then(() => {
+        expect(optedOutSpy).toHaveBeenCalledWith(jasmine.any(Error));
+      });
+    });
+
+    it("considers the user opted in after the user opts out then opts in", () => {
+      const optedInSpy = jasmine.createSpy();
+      optIn.enable(cookie);
+      optIn.setPurposes("none");
+      optIn.setPurposes("all");
+      optIn.whenOptedIn().then(optedInSpy);
+
+      expect(cookie.set).toHaveBeenCalledWith("optIn", "all");
+      expect(optIn.isOptedIn()).toBe(true);
+      return flushPromises().then(() => {
+        expect(optedInSpy).toHaveBeenCalled();
+      });
+    });
+
+    it("considers the user opted out after the user opts in then opts out", () => {
+      const optedOutSpy = jasmine.createSpy();
+      optIn.enable(cookie);
+      optIn.setPurposes("all");
+      optIn.setPurposes("none");
+      optIn.whenOptedIn().catch(optedOutSpy);
+
+      expect(cookie.set).toHaveBeenCalledWith("optIn", "none");
+      expect(optIn.isOptedIn()).toBe(false);
+      return flushPromises().then(() => {
+        expect(optedOutSpy).toHaveBeenCalledWith(jasmine.any(Error));
+      });
+    });
+  });
+});
```

Comment:
Another test case would be to do a nested whenOptedIn() call.  Although, looking at the code, I think it would pass.

Reply:
I added another test that I think addresses what you're talking about.

Reply:
Looks good.  Thanks.